### ISSUE TITLE
Add course and/or project statistics page

### ISSUE THREAD

**Issue by james-perretta** [eecs-autograder/autograder.io#4]
- Descriptive stats
- Score histogram
- Submissions over time
- Number of pass/fail per test case

**Comment by lslavice** [comment]
PICTURES!!!!! :) More ideas (in no particular order):

- histogram of times of first submission (ideally in terms of "days-hours before due day/time")
- histogram of times of first full-points submission (ideally in terms of "days-hours before due day/time")
- student "dashboard" showing a graph of when they submitted for each project (ideally in terms of "days-hours before due day/time") and what the score was (e.g. type in a uniqname and it brings up this visual); if you ever get to this one, let me know and I can sketch out what I mean
- percentage of students (NOT including staff) that have completed the project so far. This might be part of your "descriptive statistics" already

If I think of more, I will add them! Thanks for asking!

**Comment by james-perretta** [comment]
@lslavice Thanks for your ideas! I have a few questions:

> Histogram of times of first submission/first full-points submission

What would the axes and buckets be for these? 

> Student "dashboard"

This does sound a little involved, but a sketch at your convenience would be helpful so that I can figure out just how involved it is. 

> Percentage of students (NOT including staff) that have completed the project

I think this is covered by the score histogram. Here's an example with random data:
![Screenshot from 2020-10-13 10-06-51](https://user-images.githubusercontent.com/4098992/95898637-5985e980-0d5d-11eb-89d4-c8c0989627c7.png)

**Comment by amirkamil** [comment]
Stuff I've computed myself:

- time of first submission vs. final grade
- project mean and median over time
- stats by group size (e.g. partnerships vs. solo students)

**Comment by james-perretta** [comment]
@amirkamil What time increment do you use for mean/median over time?

**Comment by amirkamil** [comment]
Whenever someone makes a new submission. Here's an example:

<img width="543" alt="Picture1" src="https://user-images.githubusercontent.com/2025676/96164181-df826b80-0ee8-11eb-9f57-cce8fd7543fb.png">

**Comment by amirkamil** [comment]
The bottom axis is by day.

**Comment by james-perretta** [comment]
@amirkamil @lslavice Here are some screenshots of what I have so far. Note that the data is random and not necessarily distributed the same way as real data would be.

## Descriptive stats and histogram
![Screenshot from 2020-10-20 11-13-45](https://user-images.githubusercontent.com/4098992/96606665-84c08980-12c5-11eb-9748-7b8d0cc993db.png)

## Number of submissions per hour
![image](https://user-images.githubusercontent.com/4098992/96606862-bc2f3600-12c5-11eb-9f8d-f1b0837321eb.png)

## Pass/Fail counts per test case
![Screenshot from 2020-10-20 11-20-51](https://user-images.githubusercontent.com/4098992/96607454-614a0e80-12c6-11eb-9340-ade87fd1d06d.png)

## Final score vs time of first submission
![image](https://user-images.githubusercontent.com/4098992/96606915-ca7d5200-12c5-11eb-94ec-67b49069dadc.png)

**Comment by amirkamil** [comment]
Looks good so far! Thanks for adding this.

**Comment by jamesjuett** [comment]
I'm really late to the party, but this will be very cool!

**Comment by aposp** [comment]
Hi @james-perretta, I've encountered an error when trying to retrieve project statistics (specifically for EECS 280 project 1)

Any of the graphs take up to 20 seconds to finish loading. When finished, no graph is shown, and the following error occurs in the console: 

![image](https://user-images.githubusercontent.com/42613926/106423755-11372800-642f-11eb-85af-0caa6425403c.png)

As far as I can tell, the choice of graph does not matter, nor does "Include Staff" being checked when trying to generate a graph. Reload Data seems to have the same effect as opening a tab for the first time as well (similar load time and results)

Something else that may be worth noting, I was able to generate these graphs somewhat recently - either Wednesday or Thursday? Friday morning is when I tried again and first encountered issues. No changes since then. My guess is that a specific submit somewhere broke the graph generation.

This is not an urgent issue or anything, just wanted to document what occurred. Let me know if you need further information, testing, etc. from me.

**Comment by james-perretta** [comment]
@aposp Thanks, I'll take a look. : )

**Comment by Faryab** [comment]
These are great additions!

Question: Is the submissions over time graph only available over the entire life of the project? I was specifically interested in submissions over the last two days or x number of days. The graph become increasingly hard to read the longer a project is released. 

Another way to solve this would maybe be to add the dates in addition to the time in each of the x-axis labels. 

<img width="1378" alt="image" src="https://user-images.githubusercontent.com/40459661/106515046-e6b0a200-64f6-11eb-85bd-d7749ba869fd.png">

Another question: Are these statistics all in EST, or do they change depending on the timezone of the user? I know the due-dates shown are in my own timezone, so I want to make sure if I share statistics with the rest of the class they are timezone representative.

**Comment by james-perretta** [comment]
@aposp It looks like the problem has to do with extensions, which makes some result objects null. I'm almost done with a fix and will probably finish it up and deploy it tomorrow.

**Comment by james-perretta** [comment]
@Faryab TL;DR I think your suggestions should be possible, I just have to find the right chart.js settings. : )

> These are great additions!
Thank you!

> I was specifically interested in submissions over the last two days or x number of days. The graph become increasingly hard to read the longer a project is released. (see example below)
I'll do some research and see if we can make the graph "zoomable"

> Maybe one way to solve this would maybe be to add the dates in addition to the time in each of the x-axis labels.
Good point, I bet there's a setting for this.

> Another question: Are these statistics all in EST, or do they change depending on the timezone of the user? If it is the latter, can we maybe choose a way to have it be in EST?
I think they're in the browser's local timezone. I'll see if chart.js supports a fixed timezone here (it probably does).

**Comment by aposp** [comment]
That makes sense - the project 1 deadline was extended between success and failure. Thank you for looking into it!

**Comment by james-perretta** [comment]
@aposp The issue should be fixed now. : )

**Comment by james-perretta** [comment]
@Faryab Quick update: I found a plugin that supports zooming and panning, and I also changed the timestamp format to include the date. There do seem to be some limitations in chart.js with respect to showing the timezone. I'll keep looking into it but can't make any promises yet.

![submissions-per-hour-zoom](https://user-images.githubusercontent.com/4098992/106773824-7f3a4580-660f-11eb-8a3f-30f21df4cc5c.gif)

If that looks good to everyone, I can deploy it in the next couple days. : )

**Comment by Faryab** [comment]
The plugin looks great! Not having the timezone will be mildly annoying because I will probably have to use a VPN / change my timezone settings but definitely not a deal-breaker.  Thanks for your work!

**Comment by james-perretta** [comment]
Sounds good! I've opened [another issue](https://github.com/eecs-autograder/autograder.io/issues/12) to continue improving on timezone-related things. Ideally that will introduce a way to specify a course-wide timezone, which we can then use when displaying these timestamps.

**Comment by awdeorio** [comment]
Love it!  Here's a minor nitpick:  When I scroll over the "Submissions Over Time" graph, the page stops scrolling and starts zooming in on that graph.  It might be nice to disable mouse-scroll-zoom on that figure.

**Comment by james-perretta** [comment]
@awdeorio I agree that that's annoying. Scroll zoom is what the chart-js-zoom plugin gives you out of the box when you enable zoom. Since there's no built-in zoom button option I'll have to look into some workarounds that let you zoom programmatically.

**Comment by james-perretta** [comment]
Just a quick note, the scroll issues have been fixed as per the changes discussed in #12
